### PR TITLE
Do final MDv3 format changes

### PR DESCRIPTION
**Pull request by akalin-keybase**
Have separate DeviceKeyInfoV3 type that uses
CryptPublicKeys instead of KIDs as key values.
This is a format change because the former is
stored as binary and the latter is stored as a
hex-encoded string.

Also tweak the names and codec params
for TLFReaderKeyBundleV3.

Add functions to BareRootMetadata to do rekeying
operations instead of manipulating the internal
maps directly.

Add TODOs for various bugs.

Add some unit tests.

### CODE REVIEW

File: libkbfs/key_bundle_v2.go [hunk 1]
```diff
@@ -0,0 +1,327 @@
+// Copyright 2016 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
```
Comment (akalin-keybase) [inline libkbfs/key_bundle_v2.go]: A lot of this file is moved from key_bundle.go

**Comment by akalin-keybase** [comment]
I will also make sure the MDv3 tests pass with this.

**Comment by strib** [comment]
I'm taking a look, but I haven't reviewed any mdv3 code before so not sure I'm the best person.  Ideally @aalness can look too.

File: libkbfs/bare_root_metadata_v3.go [hunk 2]
```diff
@@ -511,6 +511,75 @@ func (md *BareRootMetadataV3) TlfHandleExtensions() (
...
+	}
+
+	wkb, rkb, ok := getKeyBundlesV3(extra)
+	if !ok {
+		return errors.New("Key bundles missing")
+	}
+
+	for _, key := range keys {
+		for _, dkim := range wkb.Keys {
+			delete(dkim, key)
```
Comment (strib) [inline libkbfs/bare_root_metadata_v3.go]: Oh, I see the answer in a `KeyManager` TODO now.  Maybe worth a comment/TODO here as well, as it does something different than v2.  The interface comment for `RevokeDevices` might need to be tweaked as well to specify more exactly what the implementer should do with historic key gens.

File: libkbfs/bare_root_metadata_v3.go [hunk 3]
```diff
@@ -511,6 +511,75 @@ func (md *BareRootMetadataV3) TlfHandleExtensions() (
...
+
+	wkb, rkb, ok := getKeyBundlesV3(extra)
+	if !ok {
+		return errors.New("Key bundles missing")
+	}
+	dkim, ok := rkb.Keys[uid]
+	if !ok {
+		return fmt.Errorf("Could not find %s in rkb", uid)
+	}
+	// TODO: This is incorrect! Fix this.
```
Comment (aalness) [inline libkbfs/bare_root_metadata_v3.go]: nit: might want to expand on why it's wrong but works anyway.

**Comment by akalin-keybase** [comment]
okay, added him, too.

File: libkbfs/key_manager.go [hunk 4]
```diff
@@ -606,6 +608,8 @@ func (km *KeyManagerStandard) Rekey(ctx context.Context, md *RootMetadata, promp
 
 		wRemoved := km.usersWithRemovedDevices(ctx, md.TlfID(), wDkim, wKeys)
 		rRemoved := km.usersWithRemovedDevices(ctx, md.TlfID(), rDkim, rKeys)
+		// TODO: This is incorrectly true if we're only
+		// promoting readers.
```
Comment (aalness) [inline libkbfs/key_manager.go]: nit: ditto. may want to reference other location that relies on this re: my other nit.

File: libkbfs/bare_root_metadata_v3.go (hunk 3 above)
Comment (akalin-keybase) [inline libkbfs/bare_root_metadata_v3.go]: Done.

File: libkbfs/key_manager.go (hunk 4 above)
Comment (akalin-keybase) [inline libkbfs/key_manager.go]: Done.

File: libkbfs/bare_root_metadata_v3.go (hunk 2 above)
Comment (akalin-keybase) [inline libkbfs/bare_root_metadata_v3.go]: I ended up scrapping this code in favor of a `RevokeRemovedDevices` function, and fixing the KeyManager bug that caused server halves to be dropped for MDv3. Doing this required writing a lot more unit tests, which I did.

**Comment by akalin-keybase** [comment]
I added a substantial amount of code so this is worth another look. However, since most of the new code is tests, and this passes all tests with MDv3 set, I'll merge this now and vendor on the server, so we can move forward with MDv3.

**Pull request merged**